### PR TITLE
[WOOMOB-3800] Use authentication endpoints for sites and WebViews

### PR DESCRIPTION
**Pull request by hichamboushaba**
Part of: WOOMOB-3800

## Description

> [!NOTE]
> **About 63% of the raw diff is tests and test support:** 878 lines (`+877/-1`), covering direct `Site` projection, endpoint identity mismatch and corruption, authenticated WebView URL selection, and the in-flight credential navigation boundary.
>
> Production is 520 raw lines (`+480/-40`), or 468 lines under Danger's test/comment/blank exclusions (`+432/-36`)—above the 300-line guideline. Keeping direct-site projection, credential-context selection, and the WebKit navigation gate together avoids an incomplete consumer path that could either forget verified endpoints or post credentials outside the validated transaction.

### Why

PRs #17718–#17724 establish the endpoint trust boundary, use it in both native authentication transports, persist verified custom endpoints, and let merchants recover relocated sign-in and dashboard addresses. The remaining direct `Site` and authenticated WebView consumers still need to use those identity-bound values.

Without this layer, a transient direct `Site` can regain standard `wp-login.php`/`wp-admin` values after fetch or restoration, and authenticated WebViews can start from the wrong login entry. WebKit credential submission also needs its own navigation boundary because browser redirects can preserve a credential POST body.

### Why separate PR

This PR contains the complete consumer layer for verified endpoint metadata while leaving the recovery contract and UI unchanged. It is independently releasable on top of #17724: standard endpoints retain their existing behavior, unusable metadata remains inert, and canonical `Site.url` identity is never rewritten.

### How

- Overlays verified login and admin endpoints on the placeholder direct `Site` at both fetch/refresh and restoration seams, including the later metadata copies, while preserving canonical `Site.url`.
- Requires the credential identity, fetched-site identity, and persisted endpoint identity to match before projecting values, with only the endpoint policy's default-port HTTP-to-HTTPS promotion allowed. Corrupt, unsafe, or mismatched metadata falls back without mutating credentials or deleting secrets.
- Resolves authenticated WebView endpoint context for the exact selected WP.org credentials, preferring explicitly supplied credentials when present and preserving standard endpoint behavior otherwise.
- Posts directly to the persisted configured login entry and uses the validated nonce destination as `redirect_to`.
- Gates the in-flight credential navigation to the initial main-frame POST and policy-approved bodyless destinations. Cross-site, downgraded, subframe, download, repeated, or excessive redirects fail closed.
- Replaces one policy-approved body-preserving destination request with a fresh bodyless GET after the WebKit policy callback returns, suppresses only its expected WebKit cancellation, and clears authentication state on completion, failure, dismissal, reload, or Web Content process termination.
- Never opens an external browser automatically when authenticated WebView credential submission fails.

WebKit deliberately does not perform the URLSession login flow's HTML preflight or transaction-local form-action discovery. It uses the persisted configured login entry and constrains every follow-up request with the navigation gate.

Deliberately excluded: recovery contract/UI changes, new endpoint persistence, authentication transport rewrites, QR or WordPress.com changes, strings, analytics, and release notes.

### Stack context

1. #17718 — endpoint trust boundary and shared rules
2. #17719 — interactive authentication
3. #17720 — runtime authentication
4. #17721 — endpoint persistence and lifecycle
5. #17724 — recovery contract and inline UI
6. This PR — direct `Site` and authenticated WebView consumers

Automated verification: all 68 focused `StoresManagerTests` and `WordPressOrgCredentialsAuthenticatorTests` passed. The WooCommerce app built successfully for an iPhone 17 Pro simulator. Changed-file SwiftLint, Swift parsing, `git diff --check`, and the changed-Swift 163-character scan also passed. The built app launched to the expected login prologue without crashing.

Latest review fix (`fab24ca86a`): matches WebKit's legacy `WebKitErrorDomain` for policy cancellation code 102. Regression tests cover cancellation before and after the replacement GET and after the successful nonce response. Changed-file SwiftLint, Swift parsing, line-length and diff checks pass. The focused test rerun is blocked before execution by existing compilation errors in `HardwareTests/CardReaderEventStripeTests.swift` (`String.multipleContactlessCardsDetected` and `String.generic`); the 68-test result above predates this fix.

## Test Steps

Use a real, publicly reachable WordPress/WooCommerce test store served over HTTPS. Sign in with the store's WordPress username and password.

### 1. Custom-endpoint WebView authentication

1. On the test store, use WP Ghost to configure a custom login address and dashboard address. Keep an existing WP Admin session open so you can restore the settings.
2. In the app, sign in using the canonical store address, complete the custom-address recovery flow, and confirm the native dashboard opens.
3. Open Menu → WooCommerce Admin.
4. Confirm the store's admin loads as the authenticated merchant without another login prompt, an automatic browser launch, or a fallback to the standard WordPress login path.
5. Force-quit and restart the app, then open Menu → WooCommerce Admin again. Confirm the same authenticated destination loads.

### 2. Standard-endpoint regression

1. Use a real HTTPS test store with the standard WordPress login and admin addresses.
2. Sign in with WordPress credentials and open Menu → WooCommerce Admin.
3. Confirm the authenticated admin loads without another login prompt or an automatic browser launch.

Application-password revocation is excluded from these acceptance steps: it exposes a pre-existing trunk issue where a failed launch-time site-info fetch is not retried after password regeneration.

## Screenshots

N/A — no UI changes.

---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

No release-note entry is included because the user-facing custom-endpoint recovery behavior is already covered by #17724; this PR completes its internal consumers without adding UI or copy.

### CODE REVIEW

**Comment by dangermattic** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is larger than 300 lines of changes. Please consider splitting it into smaller PRs for easier and faster reviews.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

**Comment by hichamboushaba** [comment]
@RafaelKayumov reminder about this PR 🙂, I want to merge the whole stack at the same.

File: WooCommerce/Classes/Authentication/AuthenticatedWebViewController.swift [hunk 1]
```diff
@@ -330,19 +531,49 @@ extension AuthenticatedWebViewController: WKNavigationDelegate {
...
+    /// WebKit's internal policy-change interruption is not exposed through `WKError.Code`.
+    static let frameLoadInterruptedByPolicyChange = 102
+}
+
+private enum SiteCredentialReplacementPurpose {
+    case continueAuthentication
+    case finishAuthentication
 }
 
 extension AuthenticatedWebViewController: WKUIDelegate {
```
Comment (RafaelKayumov) [inline WooCommerce/Classes/Authentication/AuthenticatedWebViewController.swift]: The domain check here can never match. `WKError.errorDomain` is `"WKErrorDomain"`, but the "Frame load interrupted" error WebKit emits after a policy `.cancel` uses the legacy `"WebKitErrorDomain"` with code 102. The SDK's `WebKit.tbd` exports both `_WKErrorDomain` and `_WebKitErrorDomain` as separate symbols; the legacy header just isn't shipped on iOS. `WebKitViewController.swift:455` checks the code only for the same reason.

So `isWebKitPolicyInterruption` is effectively dead, and only the `NSURLErrorCancelled` branch (from `stopLoading()`) ever suppresses. Both cancel sites in this PR produce the 102 error: cancelling in `decidePolicyFor navigationResponse` (the nonce response on a successful login) and cancelling in `decidePolicyFor navigationAction` during a 307/308 redirect. Both go through WebKit's `stopLoadingForPolicyChange`, which reports `interruptedForPolicyChangeError`.

What happens next depends on whether the 102 lands before or after the `main.async` replacement closure:

- **Closure runs first** (likely, since the 102 needs an IPC round trip): the gate is already nil, so `failSiteCredentialAuthenticationAndLoadInitialURL` is a no-op, but `wasAuthenticatingWithSiteCredentials` is false and `viewModel.didFailProvisionalNavigation` fires with the spurious 102. Every caller with an `errorHandler` sees a failure on the happy path. Today that's only `GoogleAdsCampaignCoordinator`, which tracks a flow error event, so the symptom is polluted Tracks data rather than anything visible. The first caller that wires `errorHandler` to an error UI or a retry state will break on every successful login.
- **102 lands first**: the gate is still set, so the failure path stops loading and cancels the scheduled replacement. That abandons the 307/308 `continueAuthentication` handoff and falls back to a plain initial-URL load.

The unit test at `WordPressOrgCredentialsAuthenticatorTests.swift:567` masks this because it builds the error with `NSError(domain: WKError.errorDomain, code: 102)`, i.e. the domain the code checks rather than the one WebKit sends.

Suggest matching `WebKitViewController` and checking the code only, or comparing against the literal `"WebKitErrorDomain"` (there's no Swift-visible declaration for it on iOS). Either way, please switch the test to `"WebKitErrorDomain"` so it fails against the current check first.

Easy to confirm at runtime: log into a self-hosted site with wp-admin credentials, add a temporary log of `(error as NSError).domain` in `didFailProvisionalNavigation`, and open any authenticated web view (Hub Menu links, or the Google Ads flow to also see the flow error event in the Tracks output).


This is the output in `AuthenticatedWebViewController.swift` line 553 breakpoint after navigating to web admin in Menu:
```
(lldb) po (error as NSError)
Error Domain=WebKitErrorDomain Code=102 "Frame load interrupted" UserInfo={NSErrorFailingURLKey=https://rafaelcommerce.mystagingwebsite.com/wp-admin/admin-ajax.php?action=rest-nonce, NSLocalizedDescription=Frame load interrupted, _WKRecoveryAttempterErrorKey=<WKReloadFrameErrorRecoveryAttempter: 0x10fdc6860>}
```

**Review by RafaelKayumov — COMMENTED** [review]
### Out of my testing

Section 2 partially fails on the first launch after revoking the application password. The app stays signed in and generates a replacement password, but the launch-time site-info fetch in `restoreWordPressSite` gets a 401 from the revoked password before regeneration finishes, and its failure branch only logs, so the placeholder Site is kept. As a result the Menu header shows "My Store / woocommerce.com" and WooCommerce Admin opens woocommerce.com instead of the store's admin. A second cold start recovers fully. Video attached. Note: this was recorded with a local, uncommitted exemption so `normalizedToHTTPS()` skips `127.0.0.1`, because the fixture cannot complete sign-in on current trunk otherwise (see separate comment).

### Demo
https://github.com/user-attachments/assets/07024a82-b85a-4f3e-930d-086269d45e95

### Other findings

**Fixture is incompatible with current trunk (WOOMOB-3746)**
Since 14676d4c40 (2026-08-18), every WordPress.org REST request goes through `normalizedToHTTPS()` in `RESTRequest`, with no port check. Against the plain-HTTP fixture the app-password POST is sent to `https://127.0.0.1:18080/wp-json/wp/v2/users/me/application-passwords` and fails with a TLS error, so the test steps cannot reach "store opens natively" on an unpatched build. Both PRs were opened around the day that commit merged, which is probably why it wasn't hit. To run the tests I added a local, uncommitted exemption so the helper skips `127.0.0.1`; everything above was verified on that build. Suggest updating the test steps or making the fixture serve HTTPS on the default port.

**Related: `Site` admin/login URLs are also rewritten to HTTPS on decode**
The same helper is applied in `Site.swift` to `adminURL` and `loginURL`. That may conflict with this stack's rule that only default-port HTTP-to-HTTPS promotion is allowed, and is worth a look beyond the fixture case.

**Heads-up outside this stack**
WOOMOB-3746 likely breaks site-credential login for any plain-HTTP store on a non-default port, since the rewrite ignores the port. Flagging for the WOOMOB-3746 author rather than this PR.

File: WooCommerce/Classes/Authentication/AuthenticatedWebViewController.swift [hunk 2]
```diff
@@ -282,22 +333,161 @@ private extension AuthenticatedWebViewController {
...
+                self.finishSiteCredentialAuthentication(preservingExpectedCancellation: true)
+                self.loadContent(url: destinationURL)
             }
         }
     }
+
+    private func shouldSuppressSiteCredentialCancellation(for navigation: WKNavigation?, error: Error) -> Bool {
+        let error = error as NSError
+        let isExpectedCancellation = error.domain == NSURLErrorDomain && error.code == NSURLErrorCancelled
+        let isWebKitPolicyInterruption = error.domain == WKError.errorDomain &&
```
Comment (RafaelKayumov) [inline WooCommerce/Classes/Authentication/AuthenticatedWebViewController.swift]: The domain check here can never match. `WKError.errorDomain` is `"WKErrorDomain"`, but the "Frame load interrupted" error WebKit emits after a policy `.cancel` uses the legacy `"WebKitErrorDomain"` with code 102. The SDK's `WebKit.tbd` exports both `_WKErrorDomain` and `_WebKitErrorDomain` as separate symbols; the legacy header just isn't shipped on iOS. `WebKitViewController.swift:455` checks the code only for the same reason.

So `isWebKitPolicyInterruption` is effectively dead, and only the `NSURLErrorCancelled` branch (from `stopLoading()`) ever suppresses. Both cancel sites in this PR produce the 102 error: cancelling in `decidePolicyFor navigationResponse` (the nonce response on a successful login) and cancelling in `decidePolicyFor navigationAction` during a 307/308 redirect. Both go through WebKit's `stopLoadingForPolicyChange`, which reports `interruptedForPolicyChangeError`.

What happens next depends on whether the 102 lands before or after the `main.async` replacement closure:

- **Closure runs first** (likely, since the 102 needs an IPC round trip): the gate is already nil, so `failSiteCredentialAuthenticationAndLoadInitialURL` is a no-op, but `wasAuthenticatingWithSiteCredentials` is false and `viewModel.didFailProvisionalNavigation` fires with the spurious 102. Every caller with an `errorHandler` sees a failure on the happy path. Today that's only `GoogleAdsCampaignCoordinator`, which tracks a flow error event, so the symptom is polluted Tracks data rather than anything visible. The first caller that wires `errorHandler` to an error UI or a retry state will break on every successful login.
- **102 lands first**: the gate is still set, so the failure path stops loading and cancels the scheduled replacement. That abandons the 307/308 `continueAuthentication` handoff and falls back to a plain initial-URL load.

The unit test at `WordPressOrgCredentialsAuthenticatorTests.swift:567` masks this because it builds the error with `NSError(domain: WKError.errorDomain, code: 102)`, i.e. the domain the code checks rather than the one WebKit sends.

Suggest matching `WebKitViewController` and checking the code only, or comparing against the literal `"WebKitErrorDomain"` (there's no Swift-visible declaration for it on iOS). Either way, please switch the test to `"WebKitErrorDomain"` so it fails against the current check first.

Easy to confirm at runtime: log into a self-hosted site with wp-admin credentials, add a temporary log of `(error as NSError).domain` in `didFailProvisionalNavigation`, and open any authenticated web view (Hub Menu links, or the Google Ads flow to also see the flow error event in the Tracks output).


This is the output in `AuthenticatedWebViewController.swift` line 553 breakpoint after navigating to web admin in Menu:
```
(lldb) po (error as NSError)
Error Domain=WebKitErrorDomain Code=102 "Frame load interrupted" UserInfo={NSErrorFailingURLKey=https://rafaelcommerce.mystagingwebsite.com/wp-admin/admin-ajax.php?action=rest-nonce, NSLocalizedDescription=Frame load interrupted, _WKRecoveryAttempterErrorKey=<WKReloadFrameErrorRecoveryAttempter: 0x10fdc6860>}
```

**Comment by hichamboushaba** [comment]
Thanks for the thorough testing, @RafaelKayumov!

I compared the section 2 failure with current trunk. The same failure path exists there: the launch-time site-info request bypasses the validation that triggers application-password regeneration/retry, and restoration only logs its failure. Other requests regenerate the password, but nothing retries the failed site fetch, leaving the placeholder until the next restart. This is a a pre-existing bug; I have reproduced it separately on trunk, I'll createa a ticket to keep track of it.

Regarding HTTPS enforcement, I had expected IP addresses to be excluded from that change. That’s a separate discussion for WOOMOB-3746. I’ve updated this PR’s test steps to use real HTTPS stores instead of the local Docker fixture.

Your WebKit error-domain finding was valid—thanks for catching it. Fixed in `fab24ca86a`, with regression coverage using the actual `WebKitErrorDomain`, including cancellation before and after the replacement load. The test rerun is currently blocked by unrelated HardwareTests compilation errors; lint and parsing pass.

File: WooCommerce/Classes/Authentication/AuthenticatedWebViewController.swift (hunk 2 above)
Comment (hichamboushaba) [reply WooCommerce/Classes/Authentication/AuthenticatedWebViewController.swift]: Fixed in fab24ca86a

**Review by RafaelKayumov — APPROVED** [review]
Thx for the update
